### DISCUSSION TITLE
Remove agent from schedule during SimultaneousActivation

### DISCUSSION THREAD

**Discussion by clarkpetri** [mesa/mesa#1390] [Q&A] [answered]
A model I originally built with RandomActivation has the following code for when an agent selects a move which is an exit:

![Screen Shot 2022-07-20 at 10 34 52 AM](https://user-images.githubusercontent.com/99350552/180009207-705f68ee-dd86-4b8f-b660-5855876c7ba2.png)

Everything works fine with RandomActivation. In a desire to explore SimultaneousActivation I split out the agent_step to include an `advance(self)` function. If I leave the model as RandomActivation everything works. However, upon running the model with SimultaneousActivation when an agent reaches an exit and should be removed I receive the following error:

![Screen Shot 2022-07-20 at 10 38 09 AM](https://user-images.githubusercontent.com/99350552/180009966-98df35a1-79eb-4d0c-b2f2-abbeeaf75c03.png)

Line 273 is the model's step function.

That key error (11, 15) makes no sense to me. It isn't the location of the agent in question, nor a position the agent started in or even occupied. Any input on understanding this error, a correction, and ultimately removing agents from the model/schedule in SimultaneousActivation is appreciated.

**Comment by rht** [comment, marked as answer]
You can make it work by adding `agent_keys = list(self._agents.keys())` after this line https://github.com/projectmesa/mesa/blob/54f76b468de722a792c08ba176ccc4334b16944e/mesa/time.py#L147.

**Reply by clarkpetri** [reply]
That works. Thank you as always. For my understanding, is changing the time.py code a simpler correction as opposed to mucking around in my model or is agent removal in a simultaneous activation scenario needing this change at presently?

**Reply by rht** [reply]
Oh, glad that it works. I think this is a Mesa bug. Also affects `StagedActivation`. A cleaner solution is to loop `for agent_key in self._agents.keys()` since Python dict nowadays has deterministic order. But needs testing just to be sure.

**Reply by rht** [reply]
Can you test if using `for agent_key in self._agents.keys()` instead of `for agent_key in agent_keys` gives the same result? (After you have seeded the random, for deterministic result)

**Reply by clarkpetri** [reply]
It did not work. Following error:

![Screen Shot 2022-07-20 at 10 25 25 PM](https://user-images.githubusercontent.com/99350552/180116514-052fac10-dd62-497c-9841-98465e1408f4.png)
:

The code you originally advised is line 149 in time.py. I made the suggest changes to lines 147 and 151 and commented out 149, receiving the above error. Code looked like:

![Screen Shot 2022-07-20 at 10 34 46 PM](https://user-images.githubusercontent.com/99350552/180117628-84c3b53b-a377-4afa-9bbb-c43b2dca1737.png)

I'm happy to keep trouble shooting and trying things for you. Of course if I implemented the suggestion incorrectly, please let me know. Because I wasn't definitively certain if you wanted both changed or not, I did try it with 147 changed but the original line 151 and received the same error (which I expected based on my understanding of what's happening).

**Reply by rht** [reply]
Yes, your implementation is what I would do myself. Looks like the only viable option is to use my original solution.

**Reply by clarkpetri** [reply]
Sounds good and thanks again for the assist.

**Reply by rht** [reply]
I made #1391. The `SimultaneousActivation` is already known to work, but not sure for the `StagedActivation` one. I wonder if you can test the latter (check the PR for the implementation). Also, `SimultaneousActivation` is the special case of `StagedActivation`.

**Reply by clarkpetri** [reply]
Yes, `StagedActivation` worked with line 205 from the PR added to my local time.py file and my model's activation updated accordingly. Also, good to know about the special case. Makes sense when I think about it.
